pkg/controller/acmechallenges/sync.go

```diff
@@ -93,19 +91,11 @@ func (c *controller) Sync(ctx context.Context, chOriginal *cmacme.Challenge) (er
 	// cert-manager has a chance to clean up resources created for the
 	// challenge.
 	//
-	// API Transition
-	// -- Until UseDomainQualifiedFinalizer is active, we add cmacme.ACMELegacyFinalizer.
-	// -- When it is active we add cmacme.ACMEDomainQualifiedFinalizer instead.
-	//
-	// -- Both finalizers are supported, the flag just controls the one we add.
-	//
 	// -- We only need to add a finalizer label if no supported finalizer label is present.
 	if finalizerRequired(ch) {
-		finalizer := cmacme.ACMELegacyFinalizer
-		if utilfeature.DefaultFeatureGate.Enabled(feature.UseDomainQualifiedFinalizer) {
-			finalizer = cmacme.ACMEDomainQualifiedFinalizer
-		}
-		ch.Finalizers = append(ch.Finalizers, finalizer)
+		ch.Finalizers = append(slices.DeleteFunc(ch.Finalizers, func(finalizer string) bool {
+			return finalizer == cmacme.ACMELegacyFinalizer
+		}), cmacme.ACMEDomainQualifiedFinalizer)
```

Comment:
I think the deletion of the legacy finalizer should be done unconditionally, and not as a "side effect" of requiring a finalizer.